fedimint-client/src/db.rs

```diff
@@ -262,3 +276,173 @@ impl_db_lookup!(
+pub async fn apply_migrations_client(
+    db: &Database,
+    kind: String,
+    target_db_version: DatabaseVersion,
+    migrations: ClientMigrationMap,
+    module_instance_id: ModuleInstanceId,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+
+    let mut global_dbtx = db.begin_transaction().await;
+    migrate_database_version(
+        &mut global_dbtx.to_ref_nc(),
+        target_db_version,
+        Some(module_instance_id),
+        kind.clone(),
+    )
+    .await?;
+
+    let disk_version = global_dbtx
+        .get_value(&DatabaseVersionKey(module_instance_id))
+        .await;
+
+    let db_version = if let Some(disk_version) = disk_version {
+        let mut current_db_version = disk_version;
+
+        if current_db_version > target_db_version {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {kind} was higher than the code database version."
+            )));
+        }
+
+        while current_db_version < target_db_version {
+            let new_states = if let Some(migration) = migrations.get(&current_db_version) {
+                info!(target: LOG_DB, "Migrating module {kind} from {current_db_version} to {target_db_version}");
+                let active_states = get_active_states(&mut global_dbtx.to_ref_nc()).await;
+                let inactive_states = get_inactive_states(&mut global_dbtx.to_ref_nc()).await;
+
+                migration(
+                    &mut global_dbtx
+                        .to_ref_with_prefix_module_id(module_instance_id)
+                        .into_nc(),
+                    module_instance_id,
+                    active_states,
+                    inactive_states,
+                )
+                .await?
+            } else {
+                warn!("Missing client db migration for version {current_db_version}");
+                None
+            };
+
+            // If the client migration returned new states, a state machine migration has
+            // occurred, and the new states need to be persisted to the database.
+            if let Some((new_active_states, new_inactive_states)) = new_states {
+                persist_active_states(&mut global_dbtx.to_ref_nc(), new_active_states).await;
+                persist_inactive_states(&mut global_dbtx.to_ref_nc(), new_inactive_states).await;
+            }
+
+            current_db_version.increment();
+            global_dbtx
+                .insert_entry(&DatabaseVersionKey(module_instance_id), &current_db_version)
+                .await;
+        }
+
+        current_db_version
+    } else {
+        target_db_version
+    };
+
+    global_dbtx.commit_tx_result().await?;
+    info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
+    Ok(())
+}
+
+/// Reads all active states from the database and returns `Vec<DynState>`.
+pub async fn get_active_states(dbtx: &mut DatabaseTransaction<'_>) -> Vec<DynState> {
```

Comment:
The fact that we can't list only states of a given module is pretty bad. Can you add a comment to `ActiveStateKey: `TODO: we would like to have a module_instance_id prefix to be able to query only states of a givem module id`?

Reply:
Sure, I can add the todo. This would unfortunately require a somewhat nasty database migration since it is not module specific, but it can be done similar to how we just did the database version.

Reply:
TODO added